File: tome-frontend/tome/services/friendship.service.ts
import { ENV } from '@/config/env';
import { apiClient } from './api-client';
import type {
  FriendRequestDTO,
  FriendshipDTO,
  PaginatedFriendships,
  PaginatedFriendRequests,
  FriendRequestStatus,
  FriendshipStatus,
  FriendshipStatusResponse,
} from '@/types/friendship';

const FRIENDSHIP_API_URL = `${ENV.USER_DATA_API_URL}/friendships`;

/**
 * Friendship API service
 * Wraps all friendship-related backend endpoints
 */
export const friendshipApi = {
  /**
   * Send a friend request to another user
   * POST /api/friendships/requests
   */
  sendFriendRequest: async (
    friendUserId: number,
    token: string
  ): Promise<FriendRequestDTO> => {
    return apiClient.authenticatedFetch<FriendRequestDTO>(
      `${FRIENDSHIP_API_URL}/requests`,
      token,
      {
        method: 'POST',
        body: JSON.stringify({ friendUserId }),
      }
    );
  },

  /**
   * Accept a friend request
   * POST /api/friendships/requests/{id}/accept
   */
  acceptFriendRequest: async (
    requestId: number,
    token: string
  ): Promise<FriendshipDTO> => {
    return apiClient.authenticatedFetch<FriendshipDTO>(
      `${FRIENDSHIP_API_URL}/requests/${requestId}/accept`,
      token,
      {
        method: 'POST',
      }
    );
  },

  /**
   * Reject a friend request
   * POST /api/friendships/requests/{id}/reject
   */
  rejectFriendRequest: async (
    requestId: number,
    token: string
  ): Promise<void> => {
    return apiClient.authenticatedFetch<void>(
      `${FRIENDSHIP_API_URL}/requests/${requestId}/reject`,
      token,
      {
        method: 'POST',
      }
    );
  },

  /**
   * Cancel an outgoing friend request
   * DELETE /api/friendships/requests/{id}
   */
  cancelFriendRequest: async (
    requestId: number,
    token: string
  ): Promise<void> => {
    return apiClient.authenticatedFetch<void>(
      `${FRIENDSHIP_API_URL}/requests/${requestId}`,
      token,
      {
        method: 'DELETE',
      }
    );
  },

  /**
   * Get friends list (paginated)
   * GET /api/friendships?page={page}&size={size}
   * GET /api/friendships?userId={userId}&page={page}&size={size}
   *
   * @param token - Authentication token
   * @param page - Page number (0-indexed)
   * @param size - Number of items per page
   * @param userId - Optional user ID to fetch another user's friends
   */
  getFriends: async (
    token: string,
    page: number = 0,
    size: number = 20,
    userId?: number
  ): Promise<PaginatedFriendships> => {
    let url = `${FRIENDSHIP_API_URL}?page=${page}&size=${size}`;
    if (userId) {
      url += `&userId=${userId}`;
    }

    return apiClient.authenticatedFetch<PaginatedFriendships>(url, token);
  },

  /**
   * Get incoming friend requests (paginated)
   * GET /api/friendships/requests/incoming?status={status}&page={page}&size={size}
   */
  getIncomingRequests: async (
    token: string,
    status?: FriendRequestStatus,
    page: number = 0,
    size: number = 20
  ): Promise<PaginatedFriendRequests> => {
    const statusParam = status ? `&status=${status.toLowerCase()}` : '';
    return apiClient.authenticatedFetch<PaginatedFriendRequests>(
      `${FRIENDSHIP_API_URL}/requests/incoming?page=${page}&size=${size}${statusParam}`,
      token
    );
  },

  /**
   * Get outgoing friend requests (paginated)
   * GET /api/friendships/requests/outgoing?status={status}&page={page}&size={size}
   */
  getOutgoingRequests: async (
    token: string,
    status?: FriendRequestStatus,
    page: number = 0,
    size: number = 20
  ): Promise<PaginatedFriendRequests> => {
    const statusParam = status ? `&status=${status.toLowerCase()}` : '';
    return apiClient.authenticatedFetch<PaginatedFriendRequests>(
      `${FRIENDSHIP_API_URL}/requests/outgoing?page=${page}&size=${size}${statusParam}`,
      token
    );
  },

  /**
   * Unfriend a user
   * DELETE /api/friendships/{friendId}
   */
  unfriend: async (friendId: number, token: string): Promise<void> => {
    return apiClient.authenticatedFetch<void>(
      `${FRIENDSHIP_API_URL}/${friendId}`,
      token,
      {
        method: 'DELETE',
      }
    );
  },

  /**
   * Helper: Get friendship status between current user and another user
   * Client-side logic checking outgoing/incoming requests and friendships
   */
  getFriendshipStatus: async (
    userId: number,
    token: string
  ): Promise<FriendshipStatusResponse> => {
    try {
      // Check for outgoing requests
      const outgoing = await friendshipApi.getOutgoingRequests(
        token,
        'PENDING',
        0,
        1
      );
      const outgoingRequest = outgoing.content.find(
        (req) => req.addresseeId === userId
      );
      if (outgoingRequest) {
        return {
          status: 'pending_sent',
          friendRequestId: outgoingRequest.id,
        };
      }

      // Check for incoming requests
      const incoming = await friendshipApi.getIncomingRequests(
        token,
        'PENDING',
        0,
        1
      );
      const incomingRequest = incoming.content.find(
        (req) => req.requesterId === userId
      );
      if (incomingRequest) {
        return {
          status: 'pending_received',
          friendRequestId: incomingRequest.id,
        };
      }

      // Check if already friends
      const friends = await friendshipApi.getFriends(token, 0, 100);
      const friendship = friends.content.find(
        (f) => f.friendId === userId
      );
      if (friendship) {
        return {
          status: 'friends',
          friendshipId: friendship.id,
        };
      }

      // No relationship found
      return { status: 'none' };
    } catch (error) {
      console.error('[Friendship] Error getting friendship status:', error);
      return { status: 'none' };
    }
  },

  /**
   * Helper: Get count of pending incoming friend requests
   * Used for badge display
   */
  getPendingRequestCount: async (token: string): Promise<number> => {
    try {
      const result = await friendshipApi.getIncomingRequests(
        token,
        'PENDING',
        0,
        1
      );
      return result.totalElements;
    } catch (error) {
      console.error(
        '[Friendship] Error getting pending request count:',
        error
      );
      return 0;
    }
  },

  /**
   * Helper: Get total friends count
   * Used for profile display
   */
  getFriendsCount: async (token: string): Promise<number> => {
    try {
      const result = await friendshipApi.getFriends(token, 0, 1);
      return result.totalElements;
    } catch (error) {
      console.error('[Friendship] Error getting friends count:', error);
      return 0;
    }
  },
};

File: tome-backend/tome-user-data/src/main/java/com/safegergis/tome_user_data/service/FriendshipService.java
package com.safegergis.tome_user_data.service;

import java.time.OffsetDateTime;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.safegergis.tome_user_data.dto.FriendRequestDTO;
import com.safegergis.tome_user_data.dto.FriendshipDTO;
import com.safegergis.tome_user_data.dto.UserSummaryDTO;
import com.safegergis.tome_user_data.enums.FriendRequestStatus;
import com.safegergis.tome_user_data.exception.DuplicateResourceException;
import com.safegergis.tome_user_data.exception.ForbiddenException;
import com.safegergis.tome_user_data.exception.ResourceNotFoundException;
import com.safegergis.tome_user_data.mapper.FriendRequestMapper;
import com.safegergis.tome_user_data.mapper.FriendshipMapper;
import com.safegergis.tome_user_data.model.FriendRequest;
import com.safegergis.tome_user_data.model.Friendship;
import com.safegergis.tome_user_data.repository.FriendRequestRepository;
import com.safegergis.tome_user_data.repository.FriendshipRepository;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

/**
 * Service layer for friendship operations
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class FriendshipService {

    private final FriendRequestRepository friendRequestRepository;
    private final FriendshipRepository friendshipRepository;
    private final UserServiceClient userServiceClient;

    /**
     * Send a friend request to another user
     */
    @Transactional
    public FriendRequestDTO sendFriendRequest(Long requesterId, Long addresseeId) {
        log.debug("User {} sending friend request to user {}", requesterId, addresseeId);

        // Validate not sending to self
        if (requesterId.equals(addresseeId)) {
            throw new IllegalArgumentException("Cannot send friend request to yourself");
        }

        // Verify addressee exists in tome-auth
        UserSummaryDTO addressee = userServiceClient.getUser(addresseeId);

        // Check for existing friendship
        if (friendshipRepository.existsBetweenUsers(requesterId, addresseeId)) {
            throw new DuplicateResourceException("Friendship",
                    "users " + requesterId + " and " + addresseeId);
        }

        // Check for existing request in either direction
        var existingRequest = friendRequestRepository
                .findExistingRequestBetweenUsers(requesterId, addresseeId);

        if (existingRequest.isPresent()) {
            FriendRequest existing = existingRequest.get();

            // If there's a pending request from the addressee to the requester
            // inform user to check their incoming requests instead
            if (existing.getRequesterId().equals(addresseeId) &&
                    existing.getAddresseeId().equals(requesterId) &&
                    existing.getStatus() == FriendRequestStatus.PENDING) {
                throw new DuplicateResourceException("FriendRequest",
                        "This user has already sent you a friend request. Check your incoming requests.");
            }

            // If there's already a pending request from requester to addressee
            if (existing.getRequesterId().equals(requesterId) &&
                    existing.getAddresseeId().equals(addresseeId) &&
                    existing.getStatus() == FriendRequestStatus.PENDING) {
                throw new DuplicateResourceException("FriendRequest",
                        "users " + requesterId + " and " + addresseeId);
            }

            // If rejected request exists, soft delete it to allow new request
            if (existing.getStatus() == FriendRequestStatus.REJECTED) {
                existing.setDeletedAt(OffsetDateTime.now());
                friendRequestRepository.save(existing);
            }
        }

        // Create new friend request
        FriendRequest request = FriendRequest.builder()
                .requesterId(requesterId)
                .addresseeId(addresseeId)
                .status(FriendRequestStatus.PENDING)
                .build();

        FriendRequest saved = friendRequestRepository.save(request);

        // Enrich with user data
        UserSummaryDTO requester = userServiceClient.getUser(requesterId);
        return FriendRequestMapper.toDTO(saved, requester, addressee);
    }

    /**
     * Accept a friend request
     */
    @Transactional
    public FriendshipDTO acceptFriendRequest(Long addresseeId, Long requestId) {
        log.debug("User {} accepting friend request {}", addresseeId, requestId);

        // Find request
        FriendRequest request = friendRequestRepository.findById(requestId)
                .orElseThrow(() -> new ResourceNotFoundException("FriendRequest", requestId));

        // Validate addressee is the recipient
        if (!request.getAddresseeId().equals(addresseeId)) {
            throw new ForbiddenException("You are not authorized to accept this friend request");
        }

        // Validate status is PENDING
        if (request.getStatus() != FriendRequestStatus.PENDING) {
            throw new IllegalArgumentException("Friend request is not in PENDING status");
        }

        // Create friendship with ordered IDs
        Friendship friendship = createFriendship(request.getRequesterId(), request.getAddresseeId());
        Friendship saved = friendshipRepository.save(friendship);

        // Soft delete the request
        request.setDeletedAt(OffsetDateTime.now());
        friendRequestRepository.save(request);

        // Enrich with user data and return from addressee's perspective
        Long friendId = request.getRequesterId();
        UserSummaryDTO friend = userServiceClient.getUser(friendId);

        return FriendshipMapper.toDTO(saved, addresseeId, friend);
    }

    /**
     * Reject a friend request
     */
    @Transactional
    public void rejectFriendRequest(Long addresseeId, Long requestId) {
        log.debug("User {} rejecting friend request {}", addresseeId, requestId);

        // Find request
        FriendRequest request = friendRequestRepository.findById(requestId)
                .orElseThrow(() -> new ResourceNotFoundException("FriendRequest", requestId));

        // Validate addressee is the recipient
        if (!request.getAddresseeId().equals(addresseeId)) {
            throw new ForbiddenException("You are not authorized to reject this friend request");
        }

        // Validate status is PENDING
        if (request.getStatus() != FriendRequestStatus.PENDING) {
            throw new IllegalArgumentException("Friend request is not in PENDING status");
        }

        // Update status to REJECTED
        request.setStatus(FriendRequestStatus.REJECTED);
        friendRequestRepository.save(request);
    }

    /**
     * Cancel an outgoing friend request
     */
    @Transactional
    public void cancelFriendRequest(Long requesterId, Long requestId) {
        log.debug("User {} canceling friend request {}", requesterId, requestId);

        // Find request
        FriendRequest request = friendRequestRepository.findById(requestId)
                .orElseThrow(() -> new ResourceNotFoundException("FriendRequest", requestId));

        // Validate requester is the sender
        if (!request.getRequesterId().equals(requesterId)) {
            throw new ForbiddenException("You are not authorized to cancel this friend request");
        }

        // Validate status is PENDING
        if (request.getStatus() != FriendRequestStatus.PENDING) {
            throw new IllegalArgumentException("Friend request is not in PENDING status");
        }

        // Soft delete the request
        request.setDeletedAt(OffsetDateTime.now());
        friendRequestRepository.save(request);
    }

    /**
     * Get friends list (paginated)
     */
    @Transactional(readOnly = true)
    public Page<FriendshipDTO> getFriends(Long userId, Pageable pageable) {
        log.debug("Fetching friends for user {} (page {})", userId, pageable.getPageNumber());

        // Query friendships
        Page<Friendship> friendships = friendshipRepository.findAllFriendships(userId, pageable);

        // Extract friend IDs for batch fetching
        Set<Long> friendIds = friendships.getContent().stream()
                .map(friendship -> {
                    // Get the other user's ID from perspective of userId
                    return friendship.getUserId().equals(userId)
                            ? friendship.getFriendId()
                            : friendship.getUserId();
                })
                .collect(Collectors.toSet());

        // Batch fetch user data
        Map<Long, UserSummaryDTO> users = userServiceClient.getUsers(friendIds);

        // Map to DTOs
        List<FriendshipDTO> dtos = friendships.getContent().stream()
                .map(friendship -> {
                    Long friendId = friendship.getUserId().equals(userId)
                            ? friendship.getFriendId()
                            : friendship.getUserId();
                    UserSummaryDTO friend = users.get(friendId);
                    return FriendshipMapper.toDTO(friendship, userId, friend);
                })
                .collect(Collectors.toList());

        return new PageImpl<>(dtos, pageable, friendships.getTotalElements());
    }

    /**
     * Get incoming friend requests (paginated)
     */
    @Transactional(readOnly = true)
    public Page<FriendRequestDTO> getIncomingRequests(
            Long userId,
            FriendRequestStatus status,
            Pageable pageable) {
        log.debug("Fetching incoming requests for user {} with status {} (page {})",
                userId, status, pageable.getPageNumber());

        // Query requests
        Page<FriendRequest> requests;
        if (status != null) {
            requests = friendRequestRepository.findIncomingRequests(userId, status, pageable);
        } else {
            requests = friendRequestRepository.findAllIncomingRequests(userId, pageable);
        }

        // Extract requester IDs for batch fetching
        Set<Long> requesterIds = requests.getContent().stream()
                .map(FriendRequest::getRequesterId)
                .collect(Collectors.toSet());

        // Add addressee ID (the current user)
        Set<Long> allUserIds = new HashSet<>(requesterIds);
        allUserIds.add(userId);

        // Batch fetch user data
        Map<Long, UserSummaryDTO> users = userServiceClient.getUsers(allUserIds);

        // Map to DTOs
        List<FriendRequestDTO> dtos = requests.getContent().stream()
                .map(request -> FriendRequestMapper.toDTO(
                        request,
                        users.get(request.getRequesterId()),
                        users.get(request.getAddresseeId())))
                .collect(Collectors.toList());

        return new PageImpl<>(dtos, pageable, requests.getTotalElements());
    }

    /**
     * Get outgoing friend requests (paginated)
     */
    @Transactional(readOnly = true)
    public Page<FriendRequestDTO> getOutgoingRequests(
            Long userId,
            FriendRequestStatus status,
            Pageable pageable) {
        log.debug("Fetching outgoing requests for user {} with status {} (page {})",
                userId, status, pageable.getPageNumber());

        // Query requests
        Page<FriendRequest> requests;
        if (status != null) {
            requests = friendRequestRepository.findOutgoingRequests(userId, status, pageable);
        } else {
            requests = friendRequestRepository.findAllOutgoingRequests(userId, pageable);
        }

        // Extract addressee IDs for batch fetching
        Set<Long> addresseeIds = requests.getContent().stream()
                .map(FriendRequest::getAddresseeId)
                .collect(Collectors.toSet());

        // Add requester ID (the current user)
        Set<Long> allUserIds = new HashSet<>(addresseeIds);
        allUserIds.add(userId);

        // Batch fetch user data
        Map<Long, UserSummaryDTO> users = userServiceClient.getUsers(allUserIds);

        // Map to DTOs
        List<FriendRequestDTO> dtos = requests.getContent().stream()
                .map(request -> FriendRequestMapper.toDTO(
                        request,
                        users.get(request.getRequesterId()),
                        users.get(request.getAddresseeId())))
                .collect(Collectors.toList());

        return new PageImpl<>(dtos, pageable, requests.getTotalElements());
    }

    /**
     * Unfriend / Remove friendship
     */
    @Transactional
    public void unfriend(Long userId, Long friendId) {
        log.debug("User {} unfriending user {}", userId, friendId);

        // Find friendship
        Friendship friendship = friendshipRepository.findBetweenUsers(userId, friendId)
                .orElseThrow(() -> new ResourceNotFoundException("Friendship",
                        "between users " + userId + " and " + friendId));

        // Soft delete
        friendship.setDeletedAt(OffsetDateTime.now());
        friendshipRepository.save(friendship);
    }

    /**
     * Get friends count for a user
     */
    @Transactional(readOnly = true)
    public Long getFriendsCount(Long userId) {
        log.debug("Fetching friends count for user {}", userId);
        return friendshipRepository.countFriends(userId);
    }

    /**
     * Get friendship status between two users
     * Returns status information including IDs needed for actions
     */
    @Transactional(readOnly = true)
    public Map<String, Object> getFriendshipStatus(Long userId, Long targetUserId) {
        log.debug("Checking friendship status between user {} and user {}", userId, targetUserId);

        // Check if they are friends
        var friendship = friendshipRepository.findBetweenUsers(userId, targetUserId);
        if (friendship.isPresent()) {
            return Map.of(
                "status", "friends",
                "friendshipId", friendship.get().getId()
            );
        }

        // Check for outgoing request (user -> targetUser)
        var outgoingRequest = friendRequestRepository.findPendingRequest(userId, targetUserId);
        if (outgoingRequest.isPresent()) {
            return Map.of(
                "status", "pending_sent",
                "friendRequestId", outgoingRequest.get().getId()
            );
        }

        // Check for incoming request (targetUser -> user)
        var incomingRequest = friendRequestRepository.findPendingRequest(targetUserId, userId);
        if (incomingRequest.isPresent()) {
            return Map.of(
                "status", "pending_received",
                "friendRequestId", incomingRequest.get().getId()
            );
        }

        // No relationship
        return Map.of("status", "none");
    }

    /**
     * Helper method to create friendship with ordered IDs
     */
    private Friendship createFriendship(Long userId1, Long userId2) {
        Long smallerId = Math.min(userId1, userId2);
        Long largerId = Math.max(userId1, userId2);

        return Friendship.builder()
                .userId(smallerId)
                .friendId(largerId)
                .build();
    }
}
